### PR TITLE
Front-end refactoring for display on small screens

### PR DESCRIPTION
**Pull request by ewels**
HTML / CSS fixes to make the Bootstrap navbar component work nicely on small screens:

---

### CODE REVIEW

**Comment by ewels** [comment]
Ok, I did some more stuff.. The workflow detail page is properly responsive now, with the content blocks shifting around to fit on varying screen widths:



---



---



---

The other problem was that the sidebar didn't really work on mobile. So I made it into an off-canvas sliding sidebar (based on [this](https://github.com/BlackrockDigital/startbootstrap-simple-sidebar) example).

The sidebar is shown by default on big screens and hidden by default on small screens. There's a new button in the header to toggle it:

image

To accomadate this new button I had to make a narrow version of the header image for tiny screens:

image

File: tower-web/src/app/modules/main/component/navbar/navbar.component.ts [hunk 1]
```diff
@@ -32,5 +32,12 @@ export class NavbarComponent implements OnInit {
     )
   }
 
+  toggleSidebar() {
+    console.log(this);
+    this.isSidebarToggled = !this.isSidebarToggled;
+    // TODO: Make this work properly instead of hacking with jQuery
```
Comment (ewels) [inline tower-web/src/app/modules/main/component/navbar/navbar.component.ts]: Couldn't figure out how to do this nicely on the plane - just toggling a class on another element that's in a different scope? This works fine but is hacky..

File: tower-web/src/styles.scss [hunk 2]
```diff
@@ -26,42 +26,74 @@ code {
...
+
+#sidebar-wrapper.toggled {
+  margin-left: 0;
 }
 
 #sidebar {
-  position: sticky;
-  top: 0;
-  flex: 0 0 25%;
+  // TODO: Make sticky-top behaviour work with navbar height
```
Comment (ewels) [inline tower-web/src/styles.scss]: Currently this hides underneath the top-nav when you scroll, I need to fix that still.

File: tower-backend/src/main/resources/application.yml [hunk 3]
```diff
@@ -3,7 +3,7 @@ tower:
     serverUrl: "${TOWER_SERVER_URL:`http://localhost:8000`}"
     contactEmail: "${TOWER_CONTACT_EMAIL:hello@tower.nf}"
     trustedEmails:
-      - '*@seqera.io'
+      - '*'
```
Comment (ewels) [inline tower-backend/src/main/resources/application.yml]: Not sure if you would rather leave this filled? I guess I should remove the bit in the readme about updating this if we leave it like this..

**Comment by ewels** [comment]
Sep-21-2019 00-41-48

File: tower-web/src/app/modules/main/component/navbar/navbar.component.ts (hunk 1 above)
Comment (ewels) [reply tower-web/src/app/modules/main/component/navbar/navbar.component.ts]: Looking at a couple of stackoverflow posts, it seems crazy complicated to do this with angular! Maybe my little one-liner jQuery hack isn't all that bad..?! 😆

Comment (ewels) [reply tower-web/src/app/modules/main/component/navbar/navbar.component.ts]: Hang on, I think this jQuery command stops the front end from building... 🤦‍♂ Have commented out in the last commit, so sidebar doesn't work now.

File: tower-backend/src/main/resources/application.yml (hunk 3 above)
Comment (pditommaso) [reply tower-backend/src/main/resources/application.yml]: I'm changing this so that any user is whited listed by default and optional `trustedEmails` can be provided to enable this feature.

File: tower-web/src/app/modules/main/component/navbar/navbar.component.ts (hunk 1 above)
Comment (pditommaso) [reply tower-web/src/app/modules/main/component/navbar/navbar.component.ts]: Ouch! 😆

**Comment by pditommaso** [comment]
> .. we ask that you sign our Contributor License Agreement before we can accept your contribution.

Hope you will appreciate you are not asked to print and sign a pdf any more 😆

File: tower-web/src/app/modules/main/component/navbar/navbar.component.ts (hunk 1 above)
Comment (ewels) [reply tower-web/src/app/modules/main/component/navbar/navbar.component.ts]: Yes, I need to fix this 😅 Will let you know when it’s ready for testing / merging..

**Comment by ewels** [comment]
Hmm, I think that there's something not right about the CLA setup..?

image

**Comment by pditommaso** [comment]
It looks veeery slow, but waiting some seconds it shows it. 
Screenshot 2019-09-22 at 22 32 53

The gist is [this](https://gist.github.com/pditommaso/68e9a605cfb17142a870217469490820) and basically it's the same CLA used by HashiCorp.

**Comment by ewels** [comment]
Ok, got it in the end I think! Had several timeout errors and stuff, but on the third attempt (on desktop) it worked :)

Still need to fix the PR before merging though, sorry. Will try to have a look on Wednesday when I'm actually in work..! 😱

**Comment by pditommaso** [comment]
No hurry! Thanks Phil.

**Comment by ewels** [comment]
Ok, I'm getting close now.. I think I have everything displaying properly, and I am toggling the `this.isSideBarShown` variable in the `navbar` component (that's where the button is located):

https://github.com/seqeralabs/nf-tower/pull/144/files#diff-886265eb8786630007bb9ad80a052b35R40-R43

However, I'm not able to get the `home` component template to see this variable. I want to set the `toggled` class here:

https://github.com/seqeralabs/nf-tower/pull/144/files#diff-3a398b38f450e42a2e1ac4f3e062ab0eR8

Help would be welcome! 🆘

**Comment by ewels** [comment]
I feel some kind of very simple `EventEmitter` like in [this SO post](https://stackoverflow.com/a/47924683/713980) must be the answer. But I can't get it to work... :/

**Comment by ewels** [comment]
Ok, I got it to work! Via the use of an `EventEmitter`, following [this tutorial](https://angularfirebase.com/lessons/sharing-data-between-angular-components-four-methods/#Child-to-Parent-Sharing-Data-via-Output-and-EventEmitter).

Ok, I think this PR is now ready to review 🎉

File: tower-web/src/app/modules/main/component/navbar/navbar.component.ts (hunk 1 above)
Comment (ewels) [reply tower-web/src/app/modules/main/component/navbar/navbar.component.ts]: Fixed! 🎉

File: tower-backend/src/main/resources/application.yml (hunk 3 above)
Comment (ewels) [reply tower-backend/src/main/resources/application.yml]: Updated PR to solve merge conflicts ✨

**Comment by ewels** [comment]
One thing: I wasn't 100% happy with my icon choice for the button that toggles the sidebar. I went for it because I thought it kind of resembled the list of workflow cards. But for this and any other style choices (or code, or anything really), I of course completely defer to you guys 🙇

**Comment by pditommaso** [comment]
it looks the footer is broken 
Screenshot 2019-10-03 at 14 36 22

**Comment by pditommaso** [comment]
The collapsible sidebar is neat! Thanks a a lot for this, however there a couple of problems to solve:

* The first looks simple: it's missing the grey border on the right of the sidebar (see the arrow in the screenshot) 
* Then we should find a better place for the collapse/expand icon. On the left of the Tower logo does not work. What if the sidebar collapse to a thin vertical bar on left containing only a small expand icon? At some point, we could even use it to render quick run updates/info.

Screenshot 2019-10-03 at 14 40 56

**Comment by ewels** [comment]
What’s wrong with the footer sorry?

Good spot with the grey border!

I did wonder if you’d say that about the new button - I didn’t love it either. I had initially thought of the vertical bar idea too before I started. The problem is that on mobile screens the thin vertical bar is either too thin to press easily or takes up a significant proportion of the screen. The button takes up no extra space and is easy to tap.

I’ll have a play with a vertical bar and see how it feels. If it is as I suspect above, how about having the button visible only on xsmall mobile phone screens? Then have the thin vertical bar on medium size up..

Phil

**Comment by pditommaso** [comment]
Hey Phil 
 
> What’s wrong with the footer sorry?

The footer should be anchored to the bottom. Have a look at https://tower.nf/login

> The problem is that on mobile screens the thin vertical bar

Yeah, this is also true but Tower is a desktop-first app. Maybe we should have a version design specifically for the mobile, at the end it would be just a matter of having a separate html template. What do you think?

**Comment by pditommaso** [comment]
Just linking as potential options to explore https://medium.com/better-programming/creating-angular-webapp-for-multiple-views-and-screen-sizes-50fe8a83c433

**Comment by ewels** [comment]
No need for separate views here I think. It’s very easy to show or hide a few different elements at different screen sizes with CSS (and bootstrap). This is what I was suggesting above - your vertical bar for desktops and my navbar button for mobile. Best of both worlds 😁

**Comment by pditommaso** [comment]
Deal 👍

**Comment by ewels** [comment]
I started playing with different options, but then I thought.. Do we actually need the ability to collapse the sidebar on big screens at all?

How about on large screens, there is no change in behaviour. On small screens, with hide the sidebar by default and add the button as it is above.

**Comment by ewels** [comment]
Ok, I think that this is a better compromise.. Let me know what you think!

**Comment by pditommaso** [comment]
See your point, but IMO this would fragment even more the layout. I would prefer to keep the collapsible sidebar even on large screens.

**Comment by ewels** [comment]
You're a hard man to please 😅 

Inspiration for potential UIs:
* [AdminLTE bootstrap dashboard theme](https://adminlte.io/themes/AdminLTE/index2.html)
* [Travis](https://travis-ci.com/) (simply shows active jobs at the bottom of the page)
* [Twitter](https://twitter.com/home) (icons + text on big screens, just icons on small page)
* [TweetDeck](https://tweetdeck.twitter.com/) (icons, collapse button at bottom)

Ok, I think my preferred implementation is like TweetDeck: to have a new button fixed at the bottom of the sidebar - `« Collapse`. Then in the collapsed state, it is wide enough to show coloured icons. These are the tick / cross / running icons for each workflow, plus a `»` to expand the sidebar again.

What do you think?

**Comment by pditommaso** [comment]
> You're a hard man to please 

I'm really sorry .. but you opened the pandora box 😆 

> Ok, I think my preferred implementation is like TweetDeck

I agree. Adding also WorkMail screenshot that uses a similar design to get extra idea. 

Screenshot 2019-10-04 at 14 54 45

Screenshot 2019-10-04 at 14 54 55

If we go for the TweetDeck/WorkMail style what about the top navbar? I was imaging to keep it and just collapse the sidebar. Thought?

**Comment by ewels** [comment]
Nice! Ok, I will have a play with refactoring that. It may take me a little while to find time again whilst I balance projects, but I'll do my best.

> If we go for the TweetDeck/WorkMail style what about the top navbar? I was imaging to keep it and just collapse the sidebar. Thought?

I was thinking we keep it..

What does WorkMail do on small screens? (ie. [Chrome Dev tools > Mobile preview](https://developers.google.com/web/tools/chrome-devtools/device-mode)).

**Comment by pditommaso** [comment]
No hurry! Thanks for your patience! 

> I was thinking we keep it..

OK 

> What does WorkMail do on small screens?

Nothing ..

Screenshot 2019-10-04 at 15 58 50

**Comment by pditommaso** [comment]
OK, let's focus this on fixing the UI on mobile and then the advanced sidebar will be implemented in a separate feature. Is this complete on your side?

**Comment by ewels** [comment]
Hah, you should have never said _”no hurry”_! 😂

To be honest, separating out the sidebar stuff from the mobile stuff would take me as long as just finishing up the sidebar I think. I’ll try to find some time to come back to this ASAP.

**Comment by pditommaso** [comment]
LOL. Anyhow, I'm saying that because we are planning to add more features and change the current behaviour, therefore it may overlap with what we were discussing here. Therefore I don't want to work on something we'll to change a lot in the near feature (unless it's something trivial!)

**Comment by ewels** [comment]
I started playing with this again today. Will hopefully have a new version with you soon. At home now so can't send screenshots, but my current idea is to:

* Have a `Collapse` button alongside the search bar (half the width of the now-narrower search input)
* When collapsed, have small square buttons for expand and search (which will both expand the sidebar)  and if possible, also have icons for each workflow (running / failed / succeeded etc) which can be clicked on without expanding the sidebar.
    * On desktop, hovering over the icon will show a [tooltip](https://getbootstrap.com/docs/4.4/components/tooltips/) with the pipeline and run names.
* Revert behaviour so that the sidebar is always shown by default on page load until hidden
    * This would then be consistent between mobile and desktop.
    * Only remaining difference will be that on mobile the main page slides and on desktop the width changes

How does this sound?

----

I had a quick play implementing a narrow Amazon Mail vertical expand bar. One that kind of looked a bit like your first Amazon Mail example:


image

I really didn't like it on mobile though - to make it big enough to be tapped, it has to be fairly wide and it doesn't hold any information. So there's just a bit empty space which seems like a waste.

Looking at your [mobile screenshot](https://github.com/seqeralabs/nf-tower/pull/144#issuecomment-538409477) again now, I guess that the far left icons remain on mobile.



This makes sense, as they're wide enough to tap and semi-informative. This is pretty much the approach I'm aiming for with the above.

**Comment by ewels** [comment]
Ok @pditommaso, now ready for review I think! New sidebar implemented:

desktop-tower-sidebar


mobile-tower-sidebar

**Comment by ewels** [comment]
Still todo:

- [ ] Fix merge conflicts
- [ ] Fix tooltips on the small sidebar buttons (I thought I had done this, but they don't work in the demo)
- [ ] Fix sticky sidebar scrolling on mobile

Hopefully this is enough for you to decide if the concept is acceptable though, these are relatively minor issues :)

**Comment by pditommaso** [comment]
Wow, at first sight it looks amazing! will check soon in detail soon. 👏

**Comment by ewels** [comment]
Great! I much prefer it. Also as it’s all contained in the sidebar the code is a lot simpler - I was able to get rid of all of the state emit / listening stuff.

**Comment by pditommaso** [comment]
> Also as it’s all contained in the sidebar the code is a lot simpler - I was able to get rid of all of the state emit / listening stuff.

I love this ! :D

**Comment by ewels** [comment]
Also, if I can figure out how to pass the currently active workflow ID as a variable to the sidebar workflow cards in order to have an `.active` class, I would like to highlight the card that corresponds to the active page...

**Comment by pditommaso** [comment]
Impressive work Phil, I like it. I think in the future we may use collapse icons to access specific functions, but for now, it looks great. 

Just one thing: what about making the large collapse button just showing the arrow to leave more space to the search box? (I can do, I should be able doing that 😂)

**Comment by pditommaso** [comment]
Also maybe the lens icons showed when the sidebar is collapsed could be removed, at the end it's just opening the sidebar as the arrow.

**Comment by ewels** [comment]
> Just one thing: what about making the large collapse button just showing the arrow to leave more space to the search box?

Totally doable, but I personally prefer having things a bit more explicit if there's room. For example, even though I've been using tower for quite a long time now I only realised yesterday that you could collapse the error log on failed runs, and that was because I clicked it by accident 😅 (I changed that into a button that also says "collapse" in this PR also).



How about as a compromise the button shrinks when the search input is focussed?

Phil

**Comment by ewels** [comment]
collapse-search-focus

In fairness, could probably hide the collapse button entirely in this case too if we want.

**Comment by ewels** [comment]
Ok, I think this is now ready for review.. I've fixed all of the outstanding issues I listed above and it seems to be working pretty well for me now on both mobile preview and desktop.

I refactored one or two bits I'd written and one or two bits I hadn't (eg. it's now `position:fixed` instead of `position: sticky` for the sidebar), plus I figured out how to highlight the active card (was actually pretty easy in the end!).

Let me know what you think 😊 

---

sidebar-highlight-active

**Comment by pditommaso** [comment]
Great, thanks Phil! I agree with the collapse in the error but still not 100% sure about the one next to the search button. I think I'm going to remove the "collapse" text to keep it a bit simpler and merge everything. Great work!

**Comment by pditommaso** [comment]
OK, I was too optimistic. The content in the Welcome/user/tokens pages is aligned to right instead of being centred. I guess you changed something in the main structured that altered them. Any clue?

**Comment by pditommaso** [comment]
Also the footer is not rendered properly 

Screenshot 2019-11-29 at 16 27 46

**Comment by ewels** [comment]
I’ll take a look 👍🏻

**Comment by ewels** [comment]
Ok, both bugs fixed 👍 Let me know if you spot anything else that looks odd..

**Comment by ewels** [comment]
Ok, and I've removed the `Collapse` text from the button now too. All done I think 👍 (unless you find more funny business)..

**Comment by pditommaso** [comment]
Looks fixed! thanks!

File: tower-web/src/app/modules/main/component/workflow-status/workflow-status.component.ts [hunk 4]
```diff
@@ -25,11 +25,8 @@ export class WorkflowStatusComponent implements OnInit {
 
   private statusesTags: string[];
 
-  statusesRows: string[][];
-
   constructor() {
     this.statusesTags = getAllTaskStatusesProgressStateTags();
-    this.statusesRows = chunk(this.statusesTags, 3)
   }
```
Comment (pditommaso) [inline tower-web/src/app/modules/main/component/workflow-status/workflow-status.component.ts]: Quick question: what's the rationale of this change. I confess I don't know what `chunk` was doing here.

File: tower-web/src/assets/variables.scss [hunk 5]
```diff
@@ -11,12 +11,10 @@ $submitted-orange: #f57f27;
...
 
 $body-background: #f8f9fa;
+$sidebar-width: 22rem;
+$sidebar-collapsed-width: 4rem;
 $sidebar-background: #f8f9fa;
 $sidebar-border: 1px solid $thin-grey;
 
-$navbar-height: 64px;
+$navbar-height: 63px;
 $font-size-root: 0.90rem;
```
Comment (pditommaso) [inline tower-web/src/assets/variables.scss]: Why the navbar is 1px smaller? #curiosity :D

File: tower-web/src/app/modules/main/component/workflow-main-tabs/workflow-main-tabs.component.html [hunk 6]
```diff
@@ -2,13 +2,13 @@
...
-        <a class="nav-link active" id="commandLine-tab" data-toggle="tab" href="#commandLine" role="tab" aria-controls="commandLine">Command line</a>
+        <a class="nav-link active" id="commandLine-tab" data-toggle="tab" href="#commandLine" role="tab" aria-controls="commandLine">Command <span class="d-none d-sm-inline">line</span></a>
       </li>
       <li class="nav-item">
         <a class="nav-link" id="parameters-tab" data-toggle="tab" href="#parameters" role="tab" aria-controls="parameters">Parameters</a>
       </li>
       <li class="nav-item">
-        <a class="nav-link" id="configuration-tab" data-toggle="tab" href="#configuration" role="tab" aria-controls="configuration">Configuration</a>
+        <a class="nav-link" id="configuration-tab" data-toggle="tab" href="#configuration" role="tab" aria-controls="configuration">Config<span class="d-none d-sm-inline">uration</span></a>
       </li>
```
Comment (pditommaso) [inline tower-web/src/app/modules/main/component/workflow-main-tabs/workflow-main-tabs.component.html]: wow, this is like a pro 🏆

**Comment by pditommaso** [comment]
Ok, can't wait to release this! let's merge it! 🎉

**Comment by ewels** [comment]
haha, yay!! 🚀 🎉

File: tower-web/src/assets/variables.scss (hunk 5 above)
Comment (ewels) [reply tower-web/src/assets/variables.scss]: I think I spotted a 1px gap between the sidebar and the nav..! That was it 🤷‍♂

File: tower-web/src/app/modules/main/component/workflow-status/workflow-status.component.ts (hunk 4 above)
Comment (ewels) [reply tower-web/src/app/modules/main/component/workflow-status/workflow-status.component.ts]: `statusesRows` was how you were organising the coloured boxes into two rows of three:
image

To make these fit whatever screen size is available, I refactored the front-end CSS so that these boxes stack into columns automatically. The same 2 rows x 3 columns on large screens, but also 3 x 2 on narrower screens and 6 x 1 on tiny screens. As I was able to do this only using CSS, I was able to delete the back-end code for `statusesRows` as it wasn't needed any more.

File: tower-web/src/app/modules/main/component/workflow-main-tabs/workflow-main-tabs.component.html (hunk 6 above)
Comment (ewels) [reply tower-web/src/app/modules/main/component/workflow-main-tabs/workflow-main-tabs.component.html]: image

**Comment by pditommaso** [comment]
OK, fixed to footer adding a nested `<div>` 07a5e68167ad93a6baf797838817be2f01ed367e. 

I've made also some other few changes [here](https://github.com/seqeralabs/nf-tower/commit/9a2b72e7d2615ed15c1b0df652107e7f8f88e4b2) and [here](https://github.com/seqeralabs/nf-tower/commit/f92005b2d3174d7691e97a3d50ead5c53b016035).

**Comment by ewels** [comment]
Nice!
